pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -442,3 +443,16 @@ def reconciliate_processes(self, trace: str):
                 os.kill(pid, signal.SIGKILL)
         shutil.rmtree(sync_dir)
         raise DeadlockDetectedException(f"DeadLock detected from rank: {self.global_rank} \n {trace}")
+
+    @contextmanager
+    def block_backward_sync(self):
+        """
+        Blocks ddp sync gradients behaviour on backwards pass.
+        This is useful for skipping sync when accumulating gradients, reducing communication overhead
+        Returns: context manager with sync behaviour off
+        """
+        if isinstance(self.model, DistributedDataParallel):
```

Comment:
fwiw, ShardedDataParallel supports this too, but we're not taking advantage of it now due to this check 😞 

this is also masked with the current inheritance structure, as `sharded` doesn't override this
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/sharded.py

ideally, splitting up this inheritance like this
```
                       Parallel
           /      /             \       \
         DDP      Sharded      FDSP      Deepspeed
```
will make these opportunities more apparent 

fyi @SeanNaren

Reply:
Hey @ananthsub,

I am a bit concerned that making all plugins subclass directly from parallel would result in lot of duplicated code and higher maintenance cost, especially for sharded.

```
                       Parallel
           /      /             \       \
         DDP      Sharded      FDSP      Deepspeed
```

Reply:
Alternatively, we could remove the `if isinstance(self.model, DistributedDataParallel):` check there.

Reply:
Removing the check for DDP would affect DeepSpeed. 

Regarding code duplication, I think if we better abstract the subprocess launch or start_processes in the DDP and DDP spawn plugins to ensure that code can be shared, would this address your concern? Are there other parts of the code you're worried about duplication. 

My concern with the inheritance we have now is if things are silently not called.